alternator/streams.cc

```diff
@@ -454,5 +454,144 @@ future<executor::request_return_type> executor::describe_stream(client_state& cl
+        case shard_iterator_type::AFTER_SEQUENCE_NUMBER: os << "AFTER_SEQUENCE_NUMBER"; break;
+        case shard_iterator_type::TRIM_HORIZON: os << "TRIM_HORIZON"; break;
+        case shard_iterator_type::LATEST: os << "LATEST"; break;
+        default: throw std::invalid_argument(std::to_string(int(type)));
+    }
+    return os;
+}
+std::istream& operator>>(std::istream& is, shard_iterator_type& type) {
+    sstring s;
+    is >> s;
+    if (s == "AT_SEQUENCE_NUMBER") {
+        type = shard_iterator_type::AT_SEQUENCE_NUMBER;
+    } else if (s == "AFTER_SEQUENCE_NUMBER") {
+        type = shard_iterator_type::AFTER_SEQUENCE_NUMBER;
+    } else if (s == "TRIM_HORIZON") {
+        type = shard_iterator_type::TRIM_HORIZON;
+    } else if (s == "LATEST") {
+        type = shard_iterator_type::LATEST;
+    } else {
+        throw std::invalid_argument(s);
+    }
+    return is;
+}
+
+struct shard_iterator {    
+    static auto constexpr marker = 'I';
+
+    shard_id shard;
+    utils::UUID threshold;
+
+    shard_iterator(shard_id i, utils::UUID th)
+        : shard(i)
+        , threshold(th)
+    {}
+    shard_iterator(utils::UUID arn, db_clock::time_point ts, cdc::stream_id i, utils::UUID th)
+        : shard(arn, ts, i)
+        , threshold(th)
+    {}
+    shard_iterator(const sstring&);
+
+    friend std::ostream& operator<<(std::ostream& os, const shard_iterator& id) {
+        return os << marker << std::hex << id.threshold
+            << ':' << id.shard
+            ;
+    }
+private:
+    shard_iterator(const sstring&, size_t);
+};
+
+shard_iterator::shard_iterator(const sstring& s) 
+    : shard_iterator(s, s.find(':'))
+{}
+
+shard_iterator::shard_iterator(const sstring& s, size_t i)
+    : shard(s.substr(i + 1))
+    , threshold(s.substr(1, i))
+{
+    if (s.at(0) != marker || i == sstring::npos) {
+        throw std::invalid_argument(s);
+    }
+}
+}
+
+template<typename ValueType>
+struct rapidjson::internal::TypeHelper<ValueType, alternator::shard_iterator>
+    : public from_string_helper<ValueType, alternator::shard_iterator>
+{};
+
+template<typename ValueType>
+struct rapidjson::internal::TypeHelper<ValueType, alternator::shard_iterator_type>
+    : public from_string_helper<ValueType, alternator::shard_iterator_type>
+{};
+
+namespace alternator {
+
+future<executor::request_return_type> executor::get_shard_iterator(client_state& client_state, service_permit permit, rjson::value request) {
+    _stats.api_operations.get_shard_iterator++;
+
+    auto type = rjson::get<shard_iterator_type>(request, "ShardIteratorType");
+    auto seq_num = rjson::get_opt<utils::UUID>(request, "SequenceNumber");
+    
+    if (type < shard_iterator_type::TRIM_HORIZON && !seq_num) {
+        throw rjson::missing_parameter("SequenceNumber");
+    }
+    if (type >= shard_iterator_type::TRIM_HORIZON && seq_num) {
+        throw rjson::invalid_parameter("SequenceNumber", "Iterator of this type should not have sequence number");
```

Comment:
Technically, invalid_pararmeter translates to a validationerror. It is just as specialization. Would you prefer it if the exception types simply changed namespace and parent? 
(One thing I really dislike is writing `throw api_error("ValidateionException", ...)` everywhere. Is should at the very least be a subtype `validation_exception`.

Reply:
I understand you dislike the throw api_error() everywhere, and yes, we could have done this as a subclass of api_error. BUT, it doesn't mean you need a subclass of rjson::error or that it needs to be in rjson.hh. It's not a JSON error at all, it is a DynamoDB API error. Yes, I know that rjson::error is translated by our server to an api_error, but we did it out of necessity (the actual rjson code might throw errors) and not because we want more of this rjson::error-throwing to happen.

Please note that for the DynamoDB API errors, you cannot just invent types - "ValidationException" is one of the few types of exceptions you can use (and which you should use in this case). You can pass different strings with this ValidationException, but you cannot really invent a new CalleException.

Reply:
As I wrote above, I've ensured to separate exception types, and also created exception overloads for the dynamo errors.